fftp.1
.TH FFTP 1:LOCAL "(version 0.1)" 
.ad b
.SH NAME
fftp  \- perform czt Fast Fourier Transform (FFT) on input data with welch's method
.SH SYNOPSIS
.B fftp
[\-0] (suppress DC component)
[-1] (force single-sided spectrum)
[\-a] (animate: plot each average as generated)
[\-b<rbw>]
[\-d<dbref>] [\-f<fmin>,<fmax>] [\-f<fcenter>@<span>] [\-h] 
[\-k<sidelobe_level>] [\-m<CZT_blocksize>]
[\-n<input_blocksize>]
[\-p<fft_pad_factor>]
[\-s] (enable spectral density mode)
[\-t] (enable pdplot headers)
[\-v] (verbose mode)
[\-w<window_type>]
[\-?] (print usage)
[filenames]
.br 
.PP
.SH DESCRIPTION
.BR Fftp\^
performs Welch's FFT averaging algorithm [1] on standard input, writing
spectrum plot to standard output.  If -m<output_block_length> or
-f<frequency range> options are specified, will optionally perform Chirp
Z Transform (CZT) [2] to allow efficient zooming into a desired frequency range. 
.br
.SS OPTIONS
.TP
.I "\-0"
Supresses DC component on each block.  This is implemented by measuring the DC
value of both the real and imaginary parts of the signal after windowing and then 
subtracting a windowed DC value from the data.  This supresses the sidelobes of
the DC component which might otherwise obscure low frequency spectral components.
.TP
.I "\-1"
Display the spectrum in one-sided form.  This is usually the desired format for
real input data.  The complex double sided spectrum is always calculated, but the 
displayed one-sided power is the sum of the positive and negative powers. Ignored in
CZT mode if the start frequency is negative, or if the input data is complex.
.TP
.I "\-a"
Animate mode.  Outputs each average as it is computed rather than waiting for
the entire ensemble to be processed.  If the output is piped to pdplot(1), the user
can watch the noise averaging as it progresses.
.TP
.I "\-b <rbw>"
Set the desired resolution bandwidth in units of Hz.  Fftp will pre-read (and save)
several lines from the input file to determine the sampling rate.  From the sampling
rate and effective noise bandwidth of the window, Fftp will set the input blocksize 
to achieve the desired resolution.  After setting the blocksize, Fftp will 
then reread the saved data and proceed normally.
Any blocksize specified with the
.I "\-n"
option will be overridden. Resolution Bandwidth herein is defined as the Effective Noise
Bandwith of the window.  For most windows it is approximately 1/2 the width of the main
lobe at 3dB bandwidth.
.TP
.I "\-d<ref>"
Adjust the 0dB reference level.  By default, the signal is plotted as RMS amplitude
using 1V as the reference level.  When the 
.I "\-d<ref>"
option is specified, the signal is plotted on a logarithmic scale.  The <ref>
argument can be either a number or one of "bm", "bv", or "bu".
.I "\-dbm" 
sets the 0dB reference to a signal that would dissipate 1mW 
into 50 ohms.
.I "\-dbv"
sets the 0dB reference to 1Volt RMS.
.I "\-dbc"
sets the 0dB reference to 1Volt RMS.
.I "\-dbu"
sets the 0dB reference to the voltage that would dissipate 1mW
into 600 ohms.
.I "\-d1e-3"
sets the 0dB reference to 1mV.
.TP
.I "\-f<fmin>,<fmax>"
Uses the Chirp-Z Transform (CZT) to compute the spectrum between the two 
specified frequencies.  
The arguments should be single numbers separated by comma with no embedded spaces.
.TP
.I "\-f<fcenter>,<span>"
Uses the Chirp-Z Transform (CZT) to compute the spectrum between the two 
frequencies: <fcenter>-<span> and <fcenter>+<fspan>.
The arguments should be single numbers separated by comma with no embedded spaces.
.TP
.I "\-h"
Enable peak hold mode.  In normal operation, fftp simply averages the
power spectra for each block to provide a final, smoothed result.  In
peak hold mode, the maximum of each block's spectra is held for each
bin.  This allows the detection of bursty signals within a long dataset
that would otherwise be averaged away. 
.TP
.I "\-k<sidelobe_level_db>"
This option specifies a kaiser window with sidelobes of a given level. The
argument is a positive number between 13.26 and 120dB.  The default sidelobe
level is 120dB.
.TP
.I "\-m<CZT_blocksize>"
Enable CZT processing with the specified number of output points.
.TP
.I "\-n<input_blocksize>"
Use blocks of <input_blocksize> for the Welch averaging algorithm.  The
default blocksize is 1024.  If the data stream has less points than 
desired , Fftp will reset the blocksize, print a message to stderr to inform the
user, and continue with a shorter analysis.
.TP
.I "\-p<fft_pad_factor>"
Specify a zero padding factor for input data.  It can be difficult to
measure the amplitude of frequencies that do not fall exactly on the FFT
bins.  By zero-padding the data prior to processing, the spectral peak is
can be easily read from the output data.  The padding factor may be fractional.
A pad of "1.0" doubles the number of computed frequency points by appending 
a sequence of zero values equal to the original block length to the data prior
to performing the FFT.
.TP
.I "\-s"
Display spectral density normalized to the power in a 1Hz bandwidth.  This is
useful when the noise floor is the term of interest.  In spectral density mode, all
the spectral peak powers are reduced by the resolution bandwidth.  In normal mode, the
noise floor is increased by the resolution bandwidth and the spectral peaks are
shown at their correct power level.
.TP
.I "\-t"
Enable pdplot(1) headers in the output.  These include x/y axis labels and resolution bandwidth
values.
.TP
.I "\-v"
Plot verbose debugging information to stderr.
.TP
.I "\-w<window_type>"
Specify the window type used.  Any unique prefix of the following
strings are accepted: (b)lackman, (han)ning, (ham)ming, (n)uttal, (f)lattop,
(k)aiser or (r)rectangular.  Blackman is the default.   If (k)aiser
is specified, the sidelobe level will default to 120dB unless changed
with the 
.I "\-k<sidelobe_level_db>"
option.
.SH EXAMPLES
In the examples below, the program pd(1) refers to the pdplot(1) plotting program
which takes ASCII x,y floating point numbers, one pair per line, and plots them in
an X11 window with autoscaled axes.  Any other ascii plot package, such as
gnuplot(1), could be substituted with minor changes to the script.
.PP
A sinusoid signal at 125Hz with a dc component of 1V can be
generated and analyzed as follows.

.DS
    awk '
    BEGIN {
      OFMT="%.12g"
      for (t=0; t<5; t+=1/1024)
        print t, 1+sqrt(2)*sin(6.2832*125*t)
    }' | fftp -1 -wv -n350 -p8 | pd
.DE

.PP
Complex input data such as the following, resulting from quadrature
mixing of a single-sideband signal, yields an asymmetrical two-sided
spectrum.

.DS
    awk '
    BEGIN {
      OFMT="%.12g"
      for (t=0; t<5; t+=1/1024)
        print t, cos(6.2832*125*t), sin(6.2832*125*t)
    }' | fftp -wv -n350 -p8 | pd
.DE

.PP
With the
.B -f
option, one can zoom in on the desired signal.  Note that the resolution
of the peak is still limited by the window length selected via the
.B -n
option.  The
.B -dbv
option is also introduced in this example to enable decibel plots.

.DS
    awk '
    BEGIN {
      OFMT="%.12g"
      for (t=0; t<5; t+=1/1024)
        print t, cos(6.2832*125*t), sin(6.2832*125*t)
    }' | fftp -wv -n350 -f50@125 -dbv | pd
.DE

.PP
The following example demonstrates the measurement of wideband noise
using the spectral density mode enabled by the option
.B -s.
The generated binary noise sequence has a root-mean-squared value
of 1 "Volt".  Assuming that the noise is distributed uniformly over
the bandwidth from zero to the Nyquist frequency of 512Hz, the noise
magnitude should be 1/sqrt(512) V/sqrt(Hz) or about 44 mV/sqrt(Hz).

.DS
    awk '
    BEGIN {
      OFMT="%.12g"
      for (t=0; t<250; t+=1/1024)
        print t, (rand()<0.5) ? -1 : 1
    }' | fftp -1s -n500 -a100 -t | pd
.DE

.PP
Use of a Kaiser data window allows one to resolve low-level spectral
features that can be obscured by the sidelobes of other windows.

.DS
    awk '
    BEGIN {
      OFMT="%.12g"
      for (t=0; t<25; t+=1/1024)
        print t, sqrt(2)*sin(800*t)+(sqrt(2)/1e5)*sin(880*t)
    }' | fftp -1 -k120 -n1024 -p8 -dbv | pd
.DE

.SH REFERENCES 
.TP
[1]
P. D. Welch, "The Use of Fast Fourier 
Transform for the Estimation of Power Spectra:...",
.I IEEE Trans. on Audio and Electroacoustics,
vol. AU-15, pp. 70-73, June 1967.  Reprinted in
.I Digital Signal Processing,
IEEE Press, pp. 335-338, New York, 1972.
.TP
[2]
L. R. Rabiner et al., "The Chirp z-Transform Algorithm",
.I IEEE Trans. on Audio and Electroacoustics,
vol. AU-17, pp. 86-92, June 1969.  Reprinted in
.I Digital Signal Processing,
IEEE Press, pp. 322-328, New York, 1972.
.TP
[3]
Simon Sirin, 
.I "A DSP algorithm for frequency analysis",
January 15, 2004, Embedded Magazine.  (the czt kernel given in this article 
was the basis of the fftp implementation).
.TP
[4]
Joshua Minor, FFT code GPL 2003,2006, http://lux.vu/applets/frequency.pde
(this is the fft code used in this implementation).
.TP
[5] 
D. Richard Brown III, 
.I "Digital Signal Processing - Kaiser Window Design for Fourier Analysis", 
http://spinlab.wpi.edu/courses/ece503_2014/12-5kaiser_window_design.pdf
(this is the code used to generate the kaiser window with a given sidelobe depth).
.TP
[6] 
Rick Walker, "Pdplot - an autoscaling x11 plot package", https://github.com/~omnister

.SH AUTHOR
Rick Walker (walker@omnisterra.com).  Fftp(1) attempts to duplicate the
command line options of the welch(1) program manpage originally written by Scott
Willingham at Hewlett-Packard Laboratories.  The core algorithms for fftp were taken
from public domain sources [3,4,5] and the remainder was written without reference to 
the original welch(1) source code. 


fftp.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <strings.h>
#include <unistd.h>	// getopt
#include <getopt.h>

#include "fftlib.h"
#include "window.h"

#define MAXBUF 256	// for fgets()

#define OPTSN_DEFAULT	1024	// default block size

double dbparse(char *dbrefstring);
int parse_fspec(char *line, double *fstart, double *fstop);
void dumparray(COMPLEX *in, int npts, char *str);

typedef struct options {
    int n;
    double p;
    int wtype;
    double wbw;
    double rbw;
    double kaiser;
    int m;
    int fflag;
    double fstart;
    double fstop;
    int czt;
    int animate;
    int pdplot;
    int nodc;
    int singlesided;
    int peakhold;
    double samplerate;
    int dbmode;
    double dbref;
    int verbose;
    int in;
    int out;
    int spect;	// spectral density mode
} OPTS;

OPTS opts;

void init_opts() {
   opts.n = 0;
   opts.p = 1.0;
   opts.spect = 0;
   opts.wtype = 0;
   opts.wbw = 1.0;
   opts.rbw = 0.0;
   opts.kaiser = 120.0;		// kaiser sidelobe depth
   opts.m = 0;                  // CZT options
   opts.fflag = 0;		// 1=>fstart set, 2, fstop set, 0=> none
   opts.fstart = 0.0;           // CZT options
   opts.fstop = 0.0;            // CZT options
   opts.czt=0;                  // default to no CZT
   opts.animate = 0;
   opts.pdplot = 0;
   opts.nodc = 0;               // null dc (!implemented)
   opts.animate = 0;
   opts.singlesided = 0;
   opts.peakhold = 0;
   opts.samplerate = 1.0;
   opts.dbmode = 0;             // -db
   opts.dbref = 1.0;            // -db(m|v|u|c|<n.n>)
   opts.verbose=0;
   opts.in=0;
   opts.out=0;
}

void massage_opts() {
    int debug=1;
    char *name;

    if (wintype_rbw(opts.wtype, &opts.wbw) != 0) {
       fprintf(stderr, "couldn't get window bandwidth\n");
       exit(1);
    }

    if (wintypetoname(opts.wtype,  &name) != 0) {
       fprintf(stderr, "couldn't get window name\n");
       exit(1);
    }

    // chicken and egg problem:
    // if the user specifies the resolution bw with -b option
    // then we need the sample rate to compute N
    // How do we know sample rate until we read in the data?
    // How can we read in until we know N?
    // Solution: create a wrapper around fgets();
    // Then we can set our wrapper to save lines
    // in a queue when in pushback mode.  We can then 
    // read a small number (say 4) lines into a buffer
    // and returns the sample rate.  
    // This allows computation of N.  When
    // readin(N,buf) gets called later, for normal processing
    // we turnoff pushback mode. The existence
    // of the saved lines is checked and these data
    // points are read in as a special case and preread
    // data is cleared.

    // if opts.rbw !=0.0 then compute opts.n
    if (opts.rbw != 0.0) {

	COMPLEX readbuf[16];
	double sampletime;
	int ccflag=0;

	xfgets_pushback(1);	// turn on pushback
	readin(4,readbuf,&sampletime,0,&ccflag);
	xfgets_pushback(0); // turn off pushback
	// xfgets_dump();

	// fprintf(stderr, "sampletime is %g\n", sampletime);
	opts.samplerate = 1.0/sampletime;

        opts.n = (int)(opts.wbw*opts.samplerate)/opts.rbw;
    }

    if (opts.m != 0 && opts.n == 0) {
       opts.n = opts.m;
    } else {
       if (opts.n == 0) {
	   opts.n = OPTSN_DEFAULT;
       }
    }

    if(!(opts.czt)) {
        opts.in = snap2((int) (((double)opts.n)*opts.p));
        opts.out = opts.in;
	opts.m = opts.in;
    } else {
        if (opts.m == 0) {
           opts.m = (int) opts.p*opts.n;
        }
	// FOO
        opts.in = opts.n*opts.p;
        // opts.in = snap2(opts.n*opts.p);
        opts.out = snap2(opts.n*opts.p+opts.m);
    }

    if(opts.verbose) {
        if (opts.czt) {
	    fprintf(stderr, "# CZT mode\n");
	    fprintf(stderr, "# opts.n input windowed data size = %d\n", opts.n);
	    fprintf(stderr, "# opts.in input padded block size = %d\n", opts.in);
	    fprintf(stderr, "# opts.out output block size = %d\n", opts.out);
	    fprintf(stderr, "# opts.m output data size = %d\n", opts.m);
	    fprintf(stderr, "# output points to user = %d\n", opts.m);
	    fprintf(stderr, "# opts.singlesided = %d\n", opts.singlesided);
	    fprintf(stderr, "# opts.fflag=%d, opts.fstart=%g, opts.fstop=%g\n",
	    	opts.fflag, opts.fstart, opts.fstop);
	} else {
	    fprintf(stderr, "# FFT mode\n");
	    fprintf(stderr, "# opts.n input windowed data size = %d\n", opts.n);
	    fprintf(stderr, "# opts.in input padded block size = %d\n", opts.in);
	    fprintf(stderr, "# opts.out output points to user = %d\n", opts.out);
	    fprintf(stderr, "# opts.m output data size = %d\n", opts.m);
        }
        fprintf(stderr, "# window type = %s\n", name);
    }
}

double mag(COMPLEX z) {
   return(sqrt(z.re*z.re + z.im*z.im));
}

void putpoint(int i, double fcorrection, double foffset, double p) {
   double corr;
   if (opts.spect) {
      corr=sqrt(opts.wbw*opts.samplerate/(double) opts.n);
   } else {
      corr=1.0;
   }
   if (opts.dbmode) {
      printf("%12g %12g\n", 
      	foffset+((double)i)*fcorrection*opts.n/((double)opts.m-1.0), 
	20*log10(p*(opts.m/opts.n)/corr)-20*log10(opts.dbref));
   } else {
      printf("%12g %12g\n", 
      	foffset+((double)i)*fcorrection*opts.n/((double)opts.m-1.0), 
	p*(opts.m/opts.n)/corr);
   }
}

double max(double x, double y) {
    if (x>y) {
   	return x;
    } else {
       return y;
    }
}

int main(int argc, char **argv)
{
    extern OPTS opts;
    int i;
    COMPLEX *in;
    COMPLEX *spec;
    COMPLEX *out;
    COMPLEX *inbuf;
    COMPLEX *cztemp;
    double *win;
    double time, val;
    int nc;
    double fcorrection;
    double davg;
    double dc;		// returns dc value of each readin block
    int done = 0;
    int opt;
    int numwins = 0;
    double x;
    double gain;
    double m;
    int complex=0;	// set positive if readin() sees complex (three column) input

    init_opts();

    while ((opt = getopt(argc, argv, "01ab:d:f:hk:m:n:p:stw:v?")) != -1) {
	switch (opt) {
	case '0':
	    opts.nodc++;	// null dc component
	    break;
	case '1':
	    opts.singlesided++;	// plot singlesided spectrum
	    break;
	case 'a':		// animate: plot each avg as it is generated
	    opts.animate++;
	    break;
        case 'b':               // resolution bandwidth
	    opts.rbw = atof(optarg);
	    break;
	case 'f':		// frequency span
	    // FIXME: parse frequency span
	    opts.fflag = parse_fspec(optarg, &opts.fstart, &opts.fstop);
	    if (opts.fflag == -1) {
	       fprintf(stderr, 
	       	"error parsing -f frequency spec: %s\n", optarg);
	       exit(1);
	    }
	    opts.czt++;
	    break;
	case 'h':		// peak hold
	    opts.peakhold++;
	    break;
	case 'k':		// kaiser sidelobe depth
	    opts.kaiser = atof(optarg);
	    if (opts.kaiser < 15.0) {
	       fprintf(stderr, 
	       	"kaiser sidelobe <%s> too low, set to 15db\n", optarg);
		opts.kaiser = 15.0;
	    }
	    if (opts.kaiser > 180.0) {
	       fprintf(stderr, 
	       	"kaiser sidelobe <%s> too high, set to 180db\n", optarg);
		opts.kaiser = 180.0;
	    }
	    opts.wtype = wintype("kaiser");	// specify window type
	    win_set_sidelobe(opts.kaiser);	// and set sidelobe level
	    break;
	case 'm':		// chirp z output length
	    opts.czt++;
	    opts.m = atoi(optarg);
	    break;
	case 'n':		// input data block length
	    opts.n = atoi(optarg);
	    break;
	case 'p':		// pad factor
	    opts.p = atof(optarg);
	    break;
	case 's':		// spectral density mode
	    opts.spect++;
	    opts.pdplot++;	// turn on labels too
	    break;
	case 't':		// pdplot headers
	    opts.pdplot++;
	    break;
	case 'w':
	    opts.wtype = wintype(optarg);	// specify window type
	    win_set_sidelobe(opts.kaiser);	// and set sidelobe level
	    break;
	case 'd':		// db reference
	    opts.dbmode++;
	    opts.dbref = dbparse(optarg);	// specify db reference 
	    break;
	case 'v':		// verbose
	    printf("setting verbose\n");
	    opts.verbose++;
	    break;
	case '?':
	default:
	    fprintf(stderr, "%s: performs Welch's FFT averaging algorithm\n", argv[0]);
	    fprintf(stderr, "on standard input, writing spectrum plot to standard output.\n");
	    fprintf(stderr, "If -m<output block length> or -f<frequency range> options are\n");
	    fprintf(stderr, "specified, will optionally perform Chirp Z transform to allow\n");
	    fprintf(stderr, "efficient zooming into a desired frequency range.\n\n");
	    fprintf(stderr, "usage: %s [opts]  < stdin > stdout\n", argv[0]);
	    fprintf(stderr, "\t-0 (suppress DC component on each block)\n");
	    fprintf(stderr, "\t-1 (plot single-sided spectrum)\n");
	    fprintf(stderr, "\t-a (animate: plot each avg as generated)\n");
	    fprintf(stderr, "\t-b <resolution bw>\n");
	    fprintf(stderr, "\t-d<dbreference>\n");
	    fprintf(stderr, "\t\tcan be -dbm, -dbv, -dbc, -dbu, -db<n.n>\n");
	    fprintf(stderr, "\t-f [<fmin>,<fmax>|<fcenter>@<span>]\n");
	    fprintf(stderr, "\t-h (enable peak hold mode)\n");
	    fprintf(stderr, "\t-k <kaiser sidelobe width> default=120db\n");
	    fprintf(stderr, "\t-m <CZT output block length>\n");
	    fprintf(stderr, "\t-n <input data block length>\n");
	    fprintf(stderr, "\t-p <fft pad factor>\n");
	    fprintf(stderr, "\t-s (enable spectral density mode)\n");
	    fprintf(stderr, "\t-t (enable pdplot headers)\n");
	    fprintf(stderr, "\t-v (verbose mode)\n");
	    fprintf(stderr, "\t-w <window_type>\n");
	    fprintf(stderr, "\t\twindow can be (b)lackman, (han)ning, (ham)ming,\n");
	    fprintf(stderr, "\t\t(n)uttal, (k)aiser or (r)rectangular\n");
	    fprintf(stderr, "\t\tdefault is (b)lackman\n");
	    fprintf(stderr, "\t-? (print usage)\n");
	    exit(1);
	}
    }

    massage_opts();

    // Set up an array to hold the data, and assign the data.
    in = new_complex(opts.in);		// buffer for fft1d
    inbuf = new_complex(opts.in);	// scratch for readin();
    
    // preparing to change windowing technique
    // first create an array to hold window coefficients:
    win = new_double(opts.n);		// buffer for holding window
    //
    // then fill array with coefficients 
    //
    window_get(win,opts.n,opts.wtype);

    //for (i=0; i<opts.n; i++) {
    //   printf("# %g %g\n", (double)i, win[i]);
    //}

    // 
    // then get data as many times as we want in array in[]
    // window_do(in,win,opts.nodc,opts.singlesided);
    // (multiplies in[] by coefficients in win[])
    //


    if (opts.czt) {	// prepare czt auxiliary array
    	cztemp = new_complex(opts.out);
    	spec = new_complex(opts.m);	// hold avg power spect
    } else {
    	spec = new_complex(opts.in);	// hold avg power spect
    }

    // FIXME: readin() fails when number of input points is less than 
    // default 1024. should reset opts.n and try again.

    numwins=0;
    int nn=opts.n;
    printf("#entering while %d\n", nn);
    while (1) {
        done=readin(&nn,inbuf,&davg,numwins,&complex);
	if (done && numwins != 0) break;
        
	printf("#readin returned %d\n", nn);

	// if we run out of points on the first window, then adjust 
	// opts.n and continue

	if (numwins==0) {
	   if (nn != opts.n) {
	     fprintf(stderr,"ran out of points before reaching %d, continuing with %d\n", opts.n, nn);
	     opts.n=nn;
	   }
	}

	// don't overwrite original data for overlapping
	// windows:
        for (i = 0; i<opts.n; i++) {
	   in[i].re = inbuf[i].re; 
	   in[i].im = inbuf[i].im; 
	}
	numwins++;

	// dumparray(in, opts.n, "this is the original readin array");

        opts.samplerate = 1.0/davg;
	// fprintf(stderr,"samplerate is %g, davg=%g\n", opts.samplerate, davg);

	// old way:
	// creates window, corrects gain, every block...
	// window(in,opts.n,opts.wtype,opts.nodc,opts.singlesided);

	// new way:
	// creates window once in win[] and just reuses it...
        window_do(in,win,opts.n,opts.nodc,opts.singlesided);

	// dumparray(in, opts.n, "this is the windowed array");

	for (i = opts.n; i< opts.in; i++) {	// zero pad
	   in[i].re=0.0; in[i].im=0.0;
	}

	// dumparray(in, opts.in, "this is the zero padded array");

	if (!opts.czt) {		// normal fft
	    out=fft_1d(in,opts.in);
	    fcorrection = opts.samplerate/(double) opts.n;
	    // dumparray(in, opts.in, "this is the fftd array");
	} else {			// czt mode
	    m = (double) opts.m;
            switch (opts.fflag) {
		case 0:
		    // ((msamples-2)/msamples)*(-Fs/2) to Fs/2
		    // ((msamples-1)/msamples)*(-Fs/2)

		    opts.fstart = ((m-2.0)/(m))*(-opts.samplerate/2.0);
		    opts.fstop = opts.samplerate/(2.0);
		    break;
	    	case 1:	
		    opts.fstop = opts.samplerate*((m-1.0)/(2.0*m));
		    break;
		case 2:
		    opts.fstart = -opts.samplerate*((m-1.0)/(2.0*m));
		    break;
		case 3:
		    // FIXME: perhaps check for out of bounds frequencies here
		default:
		    break;
	    }
	    // fcorrection = opts.samplerate/m;
	    fcorrection = opts.p*(opts.fstop-opts.fstart)/m;

	    // fprintf(stderr,"fs=%g fe=%g fd=%g\n", opts.fstart, opts.fstop, opts.samplerate);
	    out=czt(in, 		//(COMPLEX*) input sequence
	            opts.in,    	//(int) size of input sequence
		    opts.m,     	//(int) size of output sequence
		    cztemp,     	//(COMPLEX*) array used to return results
		    opts.out,   	//(int) size of scratch array
		    opts.fstart,	//(double) start frequency
		    opts.fstop, 	//(double) stop frequency
		    opts.samplerate); 	//(double) sampling frequency
	}
	
	for (i = 0; i < opts.m; i++) {	
	    x = (mag(out[i])/(double)opts.m);
	    if (numwins == 1) {
		spec[i].re = x*x; 	// welch averages square of magnitude
		spec[i].im = 0.0;
	    } else {
	        if (opts.peakhold) { 		// save peaks
	       	    spec[i].re = max(spec[i].re, x*x); 
		    spec[i].im = 0.0;
	        } else {			// average sqr mags
		    spec[i].re += (x*x-spec[i].re)/((double)numwins);
		    spec[i].im = 0.0;
		}
	    }
	    if (opts.animate) {		// for now plots only half the spectrum
		putpoint(i, fcorrection, 0.0, sqrt(spec[i].re));
	    	// printf("  %12f  %12f\n", ((double) i)*fcorrection, spec[i].re);
	    }
	}
	// dumparray(spec, opts.m, "this is spec array");
    }

    if (opts.pdplot) {
	if (opts.dbmode) {
	    if (opts.spect) {
		printf("yscale 1 Magnitude (dBV/Hz)\n");
	    } else {
		printf("yscale 1 Magnitude dBv\n");
	    }
	} else {
	    if (opts.spect) {
		printf("yscale 1 Magnitude (Vrms/sqrt(Hz))\n");
	    } else {
		printf("yscale 1 Magnitude\n");
	    }
	}
        printf("xscale 1 Frequency   RBW=%g HZ  (%g dB)\n",
	        opts.wbw*opts.samplerate/(double) opts.n,
		10*log10(opts.wbw*fcorrection));
        // printf("title fs=%g rbw=%gHz, n=%d frames=%d\n",
        //	opts.samplerate, opts.wbw*fcorrection,opts.n,numwins);
    }

    if (!opts.czt && opts.singlesided && !complex) {
	double posmag, negmag, corr;
	for (i = 0; i < ((opts.m/2)+1); i++) {

	    // FOO
	    corr = 1.0; //sqrt(2.0);
	    corr = 1.0/sqrt(2.0);

	    if (i == 0) {		
		posmag = sqrt(spec[i].re);
		negmag = posmag;
	    } else {
		posmag = sqrt(spec[i].re);
		negmag = sqrt(spec[opts.m - i].re);
	    }

	    // if we are in nodc mode, the dc component is
	    // nulled out to about -350dB.  If we plot this
	    // point, it distorts the yaxis in decibal mode
	    // and makes it difficult to read the plot.  
	    // So if we are in dbmode, and nodc mode, we
	    // suppress the zero Hz component.

	    if (!(i==0 && opts.nodc && opts.dbmode)) {
	        putpoint(i, fcorrection, 0.0,
	    	    corr*(negmag + posmag));
	    }
	}
    } else {	// doublesided or czt
	if (!opts.czt) { // double sided
	    for (i = opts.m/2; i > 0; i--) {
		putpoint(-i, fcorrection, 0.0, sqrt(spec[opts.m - i].re));
	    }
	    // printf("pen\n");
	    for (i = 0; i < opts.m/2; i++) {
		putpoint(i, fcorrection, 0.0, sqrt(spec[i].re));
	    }
	} else { // czt
	    for (i = 0; i < opts.m; i++) {
		putpoint(i, fcorrection, opts.fstart, sqrt(spec[i].re));
	    }
	}
    }

    return(0);
}

//------------------------------------------------------------
// xfgets() is a replacement for fgets() with pushback.
// if xfgets_pushback(mode) is called with mode=1, then 
// lines are read from fp and saved in a linked list queue.
// if xfgets_pushback(0), the lines are read from the
// queue (in original order) until the queue is empty, and 
// then lines are read from fp again.

typedef struct strlist {
    char *s;
    struct strlist *next;
    struct strlist *prev;
} STRLIST;

static int pushback = 0;
static STRLIST *head = NULL;
static STRLIST *tail = NULL;

xfgets_pushback(int mode) {
   pushback=mode;
}

xfgets_dump() {
   STRLIST *p;
   for (p=head; p!=NULL; p=p->next) {
      fprintf(stderr,"%s\n", p->s);
   }
}

xfgets_save(char *s) {
   char *savebuf;
   STRLIST *strlist;
   STRLIST *tmp;

   // copy the string
   savebuf = malloc(strlen(s)+1);
   strcpy(savebuf, s);

   // create a STRLIST object
   strlist = malloc(sizeof(STRLIST));
   strlist -> next =  NULL;
   strlist -> prev =  NULL;
   strlist -> s = savebuf;

   if (head == NULL) {
      head = strlist;
      tail = strlist;
   } else {
      tail->next = strlist;
      strlist->prev = tail;
      tail = strlist;
   }
}

char *xfgets(char *buf, int size, FILE *fp) {
   char *s;
   STRLIST *tmp;

   if (pushback) {
       s=fgets(buf, MAXBUF, stdin);
       xfgets_save(s);
   } else {
       if (head != NULL) {
	   strcpy(buf,head->s);
           s = buf;
	   tmp = head;
	   head = head->next;
	   free(tmp);
       } else {
	   s=fgets(buf, MAXBUF, stdin);
       }
       return(s);
   }

   return(s);
}

//------------------------------------------------------------

// read in <n> doubles to array <in> return nonzero on error if eof
// computes running average of sample rate and returns
// average sample interval in <*davg>.  Gives error on
// stderr if any time spacing varies from the running average
// by more than TIMETOL.
// *complex is a pointer to a flag that reports it readin 
// ever seen a complex number on the input

#define TIMETOL 1e-1

int readin(int *nn, COMPLEX * in, double *davg, int numwins, int *complex) {
    int i;
    double t, told, delta,  re, im;
    int error=0;
    int done=0;
    static int line=0;
    char buf[MAXBUF];
    int retval;
    int offset;
    int n=*nn;

    offset=0;
    printf("foo: in readin\n");
    if (numwins != 0) {	// not on first one, so shift and overlap
        for (i = 0; i <= n/2; i++) {
	   in[i].re=in[i+n/2].re;
	   in[i].im=in[i+n/2].im;
        } 
        offset=(n%2)+n/2;
    } else {
        offset=0;
    }


    for (i = 0; i < n-offset; i++) {
	told=t;

	// read complete lines and parses them
	// so that we accept 2 and three doubles per line
	// for complex data without slipping sync

	if (xfgets(buf, MAXBUF, stdin) == NULL) {
	    done++;
	    break;
	}
	line++;

	retval = sscanf(buf,"%lf %lf %lf", &t, &re, &im);

        if (retval==2) {
	    in[i+offset].re = re;
	    in[i+offset].im = 0.0;
	} else if (retval==3) {
	    in[i+offset].re = re;
	    in[i+offset].im = im;
	    *complex=1;
	} else {
	    fprintf(stderr, "line %d: bad fmt\n", line);
	    done++;
	    break;
	}

	delta=t-told;
	if (i==2) {
	   *davg = delta;
	} else {
           *davg += (delta-*davg)/((double)i);
	}

	// printf("## %g %g\n",delta, *davg);

	if (error == 0 && 
		n > 2 && 
		fabs((delta-*davg)/(delta+*davg)) > TIMETOL) {
            error++;		// suppress second error
     	    fprintf(stderr,
	    	"# WARNING: bad timestep at line %d, %g %g %g %g\n",
		line, t, told, delta, *davg);
	}  else { 
	    error=0;
        }
    }
    printf("foo: exiting readin with %d\n",i);
    if (numwins==0) *nn = i;
    return(done);
}

/*
        -d ref By default, the spectrum is plotted in rms magnitude,
        with the pretense that the signal has the dimension of Volts.
        The -d option directs welch to output the power spectrum in
        decibels.  The argument ref can be the strings "bv", "bm", "bu",
        or "bc", or a number may be given, specifying output in dBV
        (rms), dBm (into 50 ohms), dBu (standard in audio electronics --
        Vref=sqrt(0.6) Vrms, the voltage that would dissi‐ pate 1mW into
        600 ohms), dBc (dB relative to the output signal's peak
        magnitude), or dB (rms) relative to the given number.  For
        example, "-dbm" plots the power in dBm and "-d1e-3" plots the
        magnitude in decibels relative to 1mV.
*/


double dbparse(char *dbrefstring) {
    double dbref;
    if (strncasecmp(dbrefstring,"bv",3) == 0) {
        dbref = 1.0;
    } else if (strncasecmp(dbrefstring,"bm",3) == 0) {
        dbref = 50e-3;
    } else if (strncasecmp(dbrefstring,"bu",3) == 0) {
        dbref = 600e-3;
    } else {
        fprintf(stderr,"# bad db reference \"d%s\", using dbv\n", optarg);
        dbref = 1.0;
    }
    // printf("called with %s, return %d\n",wname, type);
    return(dbref);
}

int parse_fspec(char *line, double *fstart, double *fstop) {
   int i;
   int nval;
   char buf[128];
   char *str1 = NULL;
   char *str2 = NULL;
   double tmp;
   int debug=0;
   int retcode = 0;

   if (debug) fprintf(stderr,"testing %s\n", line);
   strcpy(buf,line);

   str1=buf; str2=NULL;
   for (i=0; buf[i]!='\0' && i<(strlen(buf)-1); i++) {
      if ((buf[i]==',') && (buf[i+1]!='\0')) {
	 buf[i]='\0';	// null terminate first line
	 str2=buf+i+1;	// set str2 to second line
	 break;
      }
   }

   if (debug) fprintf(stderr,"str1=<%s>, str2=<%s>\n", str1, str2);

   if (str1 != NULL && strlen(str1) > 0) {
       nval=sscanf(str1, "%lf", &tmp);
       if (nval==1) {
          *fstart = tmp; 
	  retcode+=1;
       } else {
	  return(-1);	// unparseable 
       }
   }
   if (str2 != NULL && strlen(str2) > 0) {
       nval=sscanf(str2, "%lf", &tmp);
       if (nval==1) {
          *fstop = tmp; 
	  retcode+=2;
       } else {
	  return(-1);	// unparseable 
       }
   }
   if (debug) fprintf(stderr,"code= %d, fstart=<%g>, fstop=<%g>\n", retcode, *fstart, *fstop);
   return (retcode);
}

void dumparray(COMPLEX *in, int npts, char *str) {
   int i;
   fprintf(stderr,"# %s\n", str);
   fprintf(stderr,"# npts=%d\n",npts);
   for (i=0; i<npts; i++) {
      fprintf(stderr,"%d %g %g\n", i, in[i].re, in[i].im);
   }
}
